src/main/java/com/yahoo/bullet/record/BulletRecord.java

```diff
@@ -213,6 +216,52 @@ public TypedObject typedGet(String field, int index, String subKey) {
         return mapValue == null ? TypedObject.NULL : new TypedObject(value.getType().getSubType().getSubType(), mapValue);
     }
 
+    /**
+     * Extracts a field as a {@link TypedObject} from the record in a custom identifier format.
+     * <br>
+     * For example, suppose a record has a map of boolean maps called "aaa". Then specifying for the identifier: <br>
+     * - "aaa" will extract that map of maps <br>
+     * - "aaa.bbb" will extract the inner map that "bbb" in "aaa" to (if it exists) <br>
+     * - "aaa.bbb.ccc" will extract the boolean that "ccc" in "bbb" in "aaa" (if it exists) <br>
+     * <br>
+     * For list fields, replace the key in the identifier format with indices instead, e.g. "list.0" or "list.0.key".
+     * If your maps have string keys that are integers,
+     * This method will also work for extract similarly for primitive maps and lists.
+     *
+     * @param identifier The non-null identifier of the field to get.
+     * @return The field or {@link TypedObject#NULL} if it does not exist or the identifier does not match the type.
+     */
+    @SuppressWarnings("unchecked")
+    public TypedObject typedExtract(String identifier) {
+        try {
+            String[] keys = identifier.split(KEY_DELIMITER, 3);
+            TypedObject object = typedGet(keys[0]);
+            Type type = object.getType();
+            Object first = object.getValue();
+            if (keys.length == 1) {
+                return object;
+            }
+            Object second;
+            Type subType = type.getSubType();
+            if (object.isMap()) {
+                second = ((Map<String, ?>) first).get(keys[1]);
+            } else if (object.isList()) {
+                second = ((List<?>) first).get(Integer.parseInt(keys[1]));
+            } else {
+                return TypedObject.NULL;
+            }
+            if (keys.length == 2) {
+                return new TypedObject(subType, second);
```

Comment:
second could be null

Reply:
Good point. I didn't add it when I changed it to Typed